src/DpopAccessToken.js

```diff
@@ -0,0 +1,161 @@
+   * @param provider.issuer {string} Provider URI
+   * @param provider.keys
+   *
+   * @param options.aud {string|Array<string>} Audience for the token
+   *   (such as the Relying Party client_id)
+   * @param options.sub {string} Subject id for the token (opaque, unique to
+   *   the issuer)
+   * @param options.scope {string} OAuth2 scope
+   *
+   * Optional:
+   * @param [options.alg] {string} Algorithm for signing the access token
+   * @param [options.jti] {string} Unique JWT id (to prevent reuse)
+   * @param [options.iat] {number} Issued at timestamp (in seconds)
+   * @param [options.max] {number} Max token lifetime in seconds
+   *
+   * @returns {JWT} Access token (JWT instance)
+   */
+  static issue (provider, options) {
+    const { issuer, keys } = provider
+
+    const { aud, sub, cnf } = options
+
+    const alg = options.alg || DEFAULT_SIG_ALGORITHM
+    const jti = options.jti || random(8)
+    const iat = options.iat || Math.floor(Date.now() / 1000)
+    const max = options.max || DEFAULT_MAX_AGE
+
+    const exp = iat + max  // token expiration
+
+    const iss = issuer
+    const key = keys.token.signing[alg].privateKey
+    const kid = keys.token.signing[alg].publicJwk.kid
+
+    
+    const header = { alg, kid }
+    const payload = { iss, aud, sub, exp, iat, jti, cnf }
+
+    const jwt = new DpopAccessToken({ header, payload, key })
+
+    return jwt
+  }
+
+  /**
+   * issue
+   */
+  static issueForRequest (request, response) {
+    const { params, code, provider, client, subject, defaultRsUri } = request
+
+    const alg = client['access_token_signed_response_alg'] || DEFAULT_SIG_ALGORITHM
+    const jti = random(8)
+    const iat = Math.floor(Date.now() / 1000)
+    let aud, sub, max, scope
+
+    // authentication request
+    if (!code) {
+      aud = client['client_id']
+      if (defaultRsUri && defaultRsUri !== client['client_id']) {
+        aud.push(defaultRsUri)
+      }
+      sub = subject['_id']
+      max = parseInt(params['max_age']) || client['default_max_age'] || DEFAULT_MAX_AGE
+      scope = request.scope
+
+    // token request
+    } else {
+      aud = [ code.aud ]
+      if (defaultRsUri && defaultRsUri !== code.aud) {
+        aud.push(defaultRsUri)
+      }
+      sub = code.sub
+      max = parseInt(code['max']) || client['default_max_age'] || DEFAULT_MAX_AGE
+      scope = code.scope
+    }
+
+    const cnf = request.dpopJwk;
+
+    const options = { aud, sub, scope, alg, jti, iat, max, cnf }
+
+    let header, payload
+
+    return Promise.resolve()
+      .then(() => DpopAccessToken.issue(provider, options))
+      .then(jwt => {
+        header = jwt.header
+        payload = jwt.payload
+
+        return jwt.encode()
+      })
+      // set the response properties
+      .then(compact => {
+        response['access_token'] = compact
+        response['token_type'] = 'Bearer'
+        response['expires_in'] = max
+      })
```

Comment:
This is not asynchronous code, what does it bring to break it up into a promise chain ?